Frontend/HsTypeChecker.hs

```diff
@@ -504,29 +505,44 @@ elabTmCase scr alts = do
   fc_rhs_ty <- liftGenM (elabMonoTy rhs_ty)
   return (rhs_ty, FcTmCaseTc fc_scr_ty fc_rhs_ty fc_scr fc_alts)
 
+type RnBinds = [(RnTmVar, RnPolyTy)]
+
+sortBinds :: RnBinds -> RnBinds
+sortBinds = sortBy (compare `on` fst)
+
 -- | Elaborate a case alternative
 elabHsAlt :: RnMonoTy         {- Type of the scrutinee  -}
           -> RnMonoTy         {- Result type            -}
           -> RnAlt            {- Case alternative       -}
           -> GenM (FcAlt 'Tc) {- Elaborated alternative -}
-elabHsAlt scr_ty res_ty (HsAlt p rhs) = notImplemented "elabHsAlt"
-  -- do
-  -- (binds, fc_p)    <- elabHsPat scr_ty p []      -- Elaborate the pattern
-  -- let (tms, tys)   = unzip binds
-  -- (rhs_ty, fc_rhs) <- extendCtxTmsM tms tys (elabTerm rhs) -- Type check the right hand side
-  -- storeEqCs        [ res_ty :~: rhs_ty ]         -- All right hand sides should be the same
-  -- return (FcAlt fc_p fc_rhs)
-
-type RnPatBinds = [(RnTmVar, RnPolyTy)]
-
-sortBinds :: RnPatBinds -> RnPatBinds
-sortBinds = sortBy (compare `on` fst)
+elabHsAlt scr_ty res_ty (HsAlt p gRs) = do
+  (binds, fc_p) <- elabHsPat scr_ty p []      -- Elaborate the pattern
+  fc_gRs        <- extendTmVars binds (mapM (elabHsGuarded res_ty) gRs) -- Type check the right hand side
+  return (FcAltTc fc_p fc_gRs)
+
+elabHsGuarded :: RnMonoTy             {- Result type -}
+              -> RnGuarded            {- Guarded rhs -}
+              -> GenM (FcGuarded 'Tc) {- Elaborated guarded rhs -}
+elabHsGuarded res_ty (HsGuarded []     rhs) = do
+  (rhs_ty, fc_rhs) <- elabTerm rhs
+  storeEqCs [res_ty :~: rhs_ty]
+  return (FcGuarded [] fc_rhs)
+elabHsGuarded res_ty (HsGuarded (g:gs) rhs) = do
+  (binds, fc_g)            <- elabHsGuard g
+  (FcGuarded fc_gs fc_rhs) <- extendTmVars binds (elabHsGuarded res_ty (HsGuarded gs rhs))
+  return (FcGuarded (fc_g : fc_gs) fc_rhs)
+
+elabHsGuard :: RnGuard -> GenM (RnBinds, FcGuard 'Tc)
+elabHsGuard (HsPatGuard p tm) = do
+  (tm_ty, fc_tm) <- elabTerm tm
+  (binds, fc_p)  <- elabHsPat tm_ty p []
```

Comment:
Here you also need to make sure that the type of the pattern matches the type of the expression.

Reply:
I think elabHsPat already does this

Reply:
You are correct.